feat: write tests and refactor scripts

vishvamsinh28 opened the pull request:
This PR sets up Jest for testing Node.js scripts, refactors few existing scripts for better readability and maintainability, and adds initial tests to ensure functionality. It also includes coverage reporting setup and updates .gitignore to exclude coverage files.

vishvamsinh28 committed:
refactored scripts, added jest and wrote test for it

vishvamsinh28 committed:
added  coverage to gitignore

anshgoyalevil committed:
fix package-lock

anshgoyalevil commented on scripts/write-json.js:
```diff
@@ -0,0 +1,13 @@
+
+        await writeFile(writePAth,JSON.stringify(jsonContent));
+    }catch(error){
+        console.log(error)
```
Use `console.error` for error logging. Update the related tests.

vishvamsinh28 replied:
done

anshgoyalevil commented on tests/utils.test.js:
```diff
@@ -0,0 +1,28 @@
+
+describe('convertToJson', () => {
+  test('should return JSON object if input is valid JSON string', () => {
+    const jsonString = '{"name": "AsyncAPI", "age": 5}';
```
Can you create an actual json file (for tests only), and read its content on the go rather than hard-coding JSON strings like this?

If `fs` module isn't working in jest, you can even create js objects and use `json.stringify()` to convert objects to JSON String.

akshatnema replied:
you can create `fixtures` to store the static data that are needed in tests.

vishvamsinh28 replied:
done

anshgoyalevil commented on tests/utils.test.js:
```diff
@@ -0,0 +1,28 @@
+  });
+
+  test('should return JavaScript object if input is valid YAML string', () => {
+    const yamlString = 'name: AsyncAPI\nage: 5';
```
Try to read actual `yaml` file here, and iff its not viable, use multiline string. Example:
```js
    const yamlString = `
      name: AsyncAPI
      age: 5
    `
```

vishvamsinh28 replied:
done

anshgoyalevil commented on tests/utils.test.js:
```diff
@@ -0,0 +1,28 @@
+
+  test('should return null if input is invalid JSON and invalid YAML', () => {
+    const invalidString = 'name: AsyncAPI, age: five';
+    const result = convertToJson(invalidString);
```
Enclose it in a try catch statement, and add an expect statement for verifying the error.

vishvamsinh28 replied:
done

anshgoyalevil commented on tests/write-json.test.js:
```diff
@@ -0,0 +1,66 @@
+  test('should read a file, convert it to JSON, and write the JSON to another file', async () => {
+    const readPath = 'config/testInput.yaml';
+    const writePath = 'config/testOutput.json';
+    const fileContent = 'name: AsyncAPI\nage: 5';
```
Use multiline strings for better readability.

vishvamsinh28 replied:
done

anshgoyalevil commented on tests/write-json.test.js:
```diff
@@ -0,0 +1,66 @@
+    await writeJSON(readPath, writePath);
+
+    expect(console.log).toHaveBeenCalledWith(error);
+  });
```
Try using try catch block to catch error instead of mocking `console.log`

vishvamsinh28 replied:
done

anshgoyalevil commented on scripts/write-json.js:
```diff
@@ -0,0 +1,13 @@
+
+        await writeFile(writePAth,JSON.stringify(jsonContent));
+    }catch(error){
+        console.log(error)
```
Instead of catching the error for all three operations, you can use nested try catch or different try catch scopes to better write tests that would allow you know which out of these three operations (`readFile`, `convertToJson`, `writeFile`) has failed.

vishvamsinh28 replied:
done

anshgoyalevil commented:
In the coverage report, there need to be other scripts as well, so that we can track which script has left, i.e., has 0 coverage.
If I am correct, There is a jest configuration to define the folder to track coverage for. Try adding that

akshatnema commented on scripts/write-json.js:
Make a `util` directory inside scripts and add these functions inside this folder. Also change the filename to a relevant one.

vishvamsinh28 replied:
done

akshatnema reviewed (commented):
I see that you are using `console.log` statements inside test to assert the values. Instead, you can refactor the functions in such a way that they return some response based on execution, and you can then test their functionalities.

vishvamsinh28 committed:
write json and test updated

vishvamsinh28 committed:
utils test updated

vishvamsinh28 committed:
write json tests updated and fixtures added

vishvamsinh28 committed:
Merge branch 'master' into jestTests

vishvamsinh28 committed:
build fix

vishvamsinh28 committed:
Merge branch 'jestTests' of https://github.com/vishvamsinh28/website into jestTests

vishvamsinh28 commented:
updated the pr @anshgoyalevil @sambhavgupta0705 @akshatnema

akshatnema commented on scripts/utils/write-json.js:
Can you have a different filename here as per camelCase naming convention and a more meaningful one?

vishvamsinh28 replied:
yes will update it

anshgoyalevil commented on tests/utils.test.js:
```diff
@@ -0,0 +1,45 @@
+      result = convertToJson(jsonString);
+    } catch (error) {
+      result = error;
+    }
```
Don't assign the `error` to `result` variable. You can directly console.error the error. If the function you are testing fails to return a valid json object if the input is a valid json string, the test would throw an error, and that's what tests are made for.

anshgoyalevil commented on tests/utils.test.js:
```diff
@@ -0,0 +1,45 @@
+      result = error;
+    }
+    expect(result).toBeNull();
+  });
```
Here, the result is always null, no matter the `convertToJson` returns a null or not. The control flow is never getting into catch block.

The catch blocks in actual functions should throw an error like this

`utils.js`
![image](https://github.com/asyncapi/website/assets/94157520/d6727915-3a45-4af1-9e15-c1d47d96d265)


After doing this, the catch block in your test would hit and the error would then be an actual error

anshgoyalevil commented on scripts/utils/write-json.js:
```diff
@@ -0,0 +1,30 @@
+        await writeFile(writePath, JSON.stringify(jsonContent));
+    } catch (error) {
+        console.error(`Error writing JSON to path ${writePath}:`, error);
+    }
```
Add the `throw error` line after `console.error` lines in all these catch blocks, and update the related tests

TRohit20 reviewed (commented):
@vishvamsinh28 This PR is part of the GSoC work, correct?

vishvamsinh28 commented:
@TRohit20 Yes

akshatnema commented:
@anshgoyalevil @sambhavgupta0705 Add `gsoc` label to all the PRs made for this repo, so that in the end, it will be easier for us to collectively show the work of mentees.

anshgoyalevil committed:
Merge branch 'master' into jestTests

anshgoyalevil commented:
@vishvamsinh28 Any updates on this PR?

vishvamsinh28 committed:
test updated

vishvamsinh28 commented:
@anshgoyalevil updated the tests

anshgoyalevil commented on tests/utils.test.js:
```diff
@@ -0,0 +1,32 @@
+    try {
+    expect(convertToJson(jsonString)).toEqual(jsonObject);
+    } catch (err) {
+      console.error(err);
```
add throw statements here also

anshgoyalevil commented on tests/utils.test.js:
```diff
@@ -0,0 +1,32 @@
+    }
+  });
+
+  test('should return null if input is invalid JSON and invalid YAML', () => {
```
```suggestion
  test('should throw an error if input is invalid JSON and invalid YAML', () => {
```

anshgoyalevil committed:
fix utils.test.js

anshgoyalevil commented:
@vishvamsinh28 I have updated the `utils.test.js`. You can watch out for where exactly the `expect` statements for test cases related to errors should go.

And, for test cases which are not supposed to fail, they shouldn't ideally be enclosed in a try catch block since the function being tested upon there is already enclosed in a try catch block, which is sufficient to throw an error and fail that specific test case.

For example,

In
```js
test('should return JSON object if input is valid JSON string', () => {
    expect(convertToJson(jsonString)).toEqual(jsonObject);
  });
```
The `convertToJson` already contains a try catch block with `throw new Error` statement. Now, suppose, the function fails to parse the jsonString, the above test case would fail (and it should fail, that what tests are supposed to do) since an error would be thrown in `convertToJson`.

Update other tests whereever applicable, and make sure there is a `throw new Error` statement along with `console.error` statement in the scripts. Similarly, make sure there is no `console.error` or `throw new Error` statement inside the tests.

Another thing is the prettier. Don't know why `.js` files are getting formatted automatically with `npm run lint:fix`. Maybe a configuration error on eslint side.
For now, you make sure the files changed in this PR are formatted. (Tip: I just press `ctrl+s` inside a specific file in vscode, and it gets formatted, maybe due to prettier extension?)

anshgoyalevil committed:
Merge branch 'master' into jestTests

anshgoyalevil committed:
Update tests/utils.test.js

akshatnema commented on tests/adopters/index.test.js:
```diff
@@ -0,0 +1,21 @@
+describe('buildAdoptersList', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
```
Is it required here?

vishvamsinh28 replied:
not really but good to have if we add more tests to it

akshatnema replied:
If it will be needed, we'll add it in the future. For now, you can remove it.

vishvamsinh28 replied:
Okay 👍

akshatnema commented on tests/write-json.test.js:
```diff
@@ -0,0 +1,91 @@
+    const error = new Error('File read error');
+
+    fs.readFile.mockRejectedValue(error);
+    const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
```
Why do we need to `mockImplementation` for console.error function?

vishvamsinh28 replied:
to prevent the output to log in the console during tests. If we log errors to the console that can clutter the output and make it harder to identify test failures when we have multiple tests

akshatnema replied:
I don't think there will be any problem with console error messages. You shouldn't mock this implementation as it will help us to detect the flow of the function, on a desired test.

vishvamsinh28 replied:
Okay 👍 I'll update it

akshatnema commented on tests/write-json.test.js:
```diff
@@ -0,0 +1,91 @@
+      `Error reading file from path ${readPath}:`,
+      error
+    );
+    consoleErrorSpy.mockRestore();
```
Do we need this as we already cleared the mocks in `beforeEach` callback?

vishvamsinh28 replied:
yes we already have it in beforeEach and it works well even if i remove it but clearAllMocks does not restore it to its original implementations so i added it

akshatnema commented on tests/utils.test.js:
```diff
@@ -0,0 +1,24 @@
+
+  test('should throw an error if input is invalid JSON and invalid YAML', () => {
+    try {
+      convertToJson(invalidString);
```
Also check that there shouldn't be any value returned from the function.

akshatnema commented on tests/write-json.test.js:
```diff
@@ -0,0 +1,91 @@
+    expect(consoleErrorSpy).toHaveBeenCalledWith(
+      `Error reading file from path ${readPath}:`,
+      error
+    );
```
@anshgoyalevil @sambhavgupta0705 Instead of checking for console messages, can't we return the function from catch block as `return Promise.reject(err);` and then check for the returned value from the function. It is more conventional way to check for errors inside functions.

anshgoyalevil replied:
Yeah. I have [suggested](https://github.com/asyncapi/website/pull/3012#issuecomment-2171047555) the use of `throw new Error` for throwing errors and use of try catch block for checking the error thrown. Also, there are many `console.error` statements inside the codebase, which are supposed to be used for debugging purpose, and should not be present in production code. IMO, we should replace all such `console.error` statements with explicit error handling.

akshatnema replied:
Yeah, @vishvamsinh28  you should refactor the code as well, like along with `console.error` log messages, the function should return the error object from the catch block as well. And you can easily detect those errors inside tests, even if you have manually stubbed the error for a certain function.

vishvamsinh28 replied:
@akshatnema yea will update it and push changes soon after fixing some errors 👍👍

vishvamsinh28 committed:
name change

vishvamsinh28 commented:
I have updated the tests for throwing errors correctly i 2 have implementations for it which one should push changes for ?
<img width="926" alt="Screenshot 2024-06-18 at 11 36 13 AM" src="https://github.com/asyncapi/website/assets/90895835/66d51727-f3d2-4332-af74-1e2b776aa65e">

<img width="802" alt="Screenshot 2024-06-18 at 11 36 38 AM" src="https://github.com/asyncapi/website/assets/90895835/1bfb2b79-39ce-46b3-9e57-0530648f94ec">

i have updated script as well
<img width="612" alt="Screenshot 2024-06-18 at 11 39 54 AM" src="https://github.com/asyncapi/website/assets/90895835/8d280cda-0324-461b-9099-a6149d42e78f">

and i have removed console.error from the script

@anshgoyalevil @sambhavgupta0705 @akshatnema

vishvamsinh28 committed:
updated tests for errors

vishvamsinh28 commented:
> Another thing is the prettier. Don't know why `.js` files are getting formatted automatically with `npm run lint:fix`. Maybe a configuration error on eslint side. For now, you make sure the files changed in this PR are formatted. (Tip: I just press `ctrl+s` inside a specific file in vscode, and it gets formatted, maybe due to prettier extension?)

i just removed prettier when we added it back to the project because it was giving me a lot of lint/prettier errors 😅 
@anshgoyalevil

anshgoyalevil committed:
Merge branch 'master' into jestTests

anshgoyalevil commented:
@vishvamsinh28 Kindly update PR title, and resolve merge conflicts

akshatnema committed:
Merge branch 'master' into jestTests

akshatnema committed:
reverted package-lock.json

anshgoyalevil committed:
fix package.json

anshgoyalevil commented:
Merging since it already got Akshat's approval 🚢

anshgoyalevil commented:
/rtm
